src/daemon/ca/status.rs

```diff
@@ -65,49 +68,130 @@ impl Default for CaStatus {
-        // are locked individually.
-        let locks = Arc::new(CaLocks::default());
-        Ok(StatusStore { store, locks })
+        let store = StatusStore { store, cache };
+        store.warm()?;
+
+        Ok(store)
     }
 
-    fn status_key(ca: &Handle) -> KeyStoreKey {
-        KeyStoreKey::scoped(ca.to_string(), "status.json".to_string())
+    /// Load existing status from disk, support the pre 0.9.5 format and silently
+    /// convert it if needed.
+    fn warm(&self) -> KrillResult<()> {
+        for scope in self.store.scopes()? {
+            if let Ok(ca) = Handle::from_str(&scope) {
+                self.convert_pre_0_9_5_full_status_if_present(&ca)?;
+                self.load_full_status(&ca)?;
+            }
+        }
+
+        Ok(())
     }
 
-    /// Returns the stored CaStatus for a CA, or a default (empty) status if it can't be found
-    pub async fn get_ca_status(&self, ca: &Handle) -> Arc<CaStatus> {
-        // Try to get it from disk, if it's missing or corrupt - get a default
-        let lock = self.locks.ca(ca).await;
-        let _ = lock.read().await;
+    /// Load current status from disk, to be used when starting up. Iif there are any
+    /// issues parsing data then default values are used - this data is not critical
+    /// so any missing, corrupted, or no longer supported data format - can be ignored.
+    /// It will get updated with new status values as Krill is running.
+    fn load_full_status(&self, ca: &Handle) -> KrillResult<()> {
+        let repo: RepoStatus = self.store.get(&Self::repo_status_key(ca))?.unwrap_or_default();
+
+        // parents
+        let mut parents = ParentStatuses::default();
+        for parent_key in self.store.keys(Some(ca.to_string()), "parents-")? {
+            let parent = parent_key.name().strip_prefix("parents-").unwrap();
+            if let Some(parent) = parent.strip_suffix(".json") {
+                if let Ok(parent) = ParentHandle::from_str(parent) {
+                    let status: ParentStatus = self
+                        .store
+                        .get(&Self::parent_status_key(ca, &parent))?
+                        .unwrap_or_default();
+
+                    parents.add(parent, status);
+                }
+            }
+        }
 
-        let status: CaStatus = self.store.get(&Self::status_key(ca)).ok().flatten().unwrap_or_default();
+        // children
+        let mut children = HashMap::new();
+        for child_key in self.store.keys(Some(ca.to_string()), "children-")? {
+            let child = child_key.name().strip_prefix("children-").unwrap();
+            if let Some(child) = child.strip_suffix(".json") {
+                if let Ok(child) = ChildHandle::from_str(child) {
+                    let status: ChildStatus = self.store.get(&Self::child_status_key(ca, &child))?.unwrap_or_default();
 
-        Arc::new(status)
+                    children.insert(child, status);
```

Comment:
The mismatch of insert here but add above is a bit strange given that the `fn add()` call above actually calls `insert()` ... maybe rename `fn add()` to `fn insert()` ?

Reply:
fair point, done